libgadget/fof.c

```diff
@@ -231,38 +231,29 @@ struct FOFPrimaryPriv {
 };
 #define FOF_PRIMARY_GET_PRIV(tw) ((struct FOFPrimaryPriv *) (tw->priv))
 
+/* Get the ultimate HEAD particle of the current tree.
+ * Read-only lockless function using atomics.*/
 static int
-HEADl(int stop, int i, TreeWalk * tw)
+HEADa(int stop, int i, const int * const Head)
 {
-    int r;
-    if (i == stop) {
-        return -1;
-    }
-//    printf("locking %d by %d in HEADl stop = %d\n", i, omp_get_thread_num(), stop);
-    lock_particle(i);
-//    printf("locked %d by %d in HEADl, next = %d\n", i, omp_get_thread_num(), FOF_PRIMARY_GET_PRIV(tw)->Head[i]);
+    int r, next;
```

Comment:
Perhaps you want next as a volatile, such that the browser does not try to re-read from Head[i]. The omp atomic read is confusing. Sounds like you just need a omp flush on next after this. But if next is declared a volatile I *think* you do not need the flush neither.

Reply:
I'm not sure - I don't know the rules for volatile well enough

Reply:
OK. After reading a bit I am no longer opposing having  atomic annotations everywhere. The standard prefers doing so, and leave the detail to the compiler.